### PR TITLE
Use weakref.proxy when saving module to internal dictionaries to not increase refcount

### LINKED ISSUE #76434: `torch.nn.Module` keeps an owning reference that prevents it from properly deleting Module
### 🐛 Describe the bug

When using method `_register_load_state_dict_pre_hook` of `torch.nn.Module` with argument `with_module=True`, the call to `functools.partial` keeps an owning reference which prevents the `Module` to be cleanly deleted when using `del module`, because there is still an owning reference.

To reproduce:

```python
import torch

def hook(self, *args, **kwargs):
    pass

module = torch.nn.Module()
module._register_load_state_dict_pre_hook(hook, with_module=True)
del module # there is still an owning reference in `module._load_state_dict_pre_hooks` hidden in closure made by `functools.partial`
```

This issue surfaced in https://github.com/PyTorchLightning/pytorch-lightning/issues/12622.

Proper workaround would probably be to change the call `functools.partial(hook, self)` to `functools.partial(hook, weakref.proxy(self))`

I can submit the PR myself if needed.

### Versions

Collecting environment information...
PyTorch version: 1.11.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: Could not collect
CMake version: version 3.23.0
Libc version: glibc-2.35

Python version: 3.9.11 (main, Apr  7 2022, 15:33:34)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.17.1-zen1-1-zen-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.112
GPU models and configuration: GPU 0: NVIDIA T1200 Laptop GPU
Nvidia driver version: 510.60.02
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.3
/usr/lib/libcudnn_adv_infer.so.8.3.3
/usr/lib/libcudnn_adv_train.so.8.3.3
/usr/lib/libcudnn_cnn_infer.so.8.3.3
/usr/lib/libcudnn_cnn_train.so.8.3.3
/usr/lib/libcudnn_ops_infer.so.8.3.3
/usr/lib/libcudnn_ops_train.so.8.3.3
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.942
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.7.0.dev0
[pip3] torch==1.11.0+cu113
[pip3] torchmetrics==0.7.3
[pip3] torchtext==0.12.0
[pip3] torchvision==0.12.0+cu113
[conda] Could not collect

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by otaj**
Fixes #76434

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1165,7 +1166,7 @@ def _call_impl(self, *input, **kwargs):
             grad_fn = var.grad_fn
             if grad_fn is not None:
                 for hook in non_full_backward_hooks:
-                    wrapper = functools.partial(hook, self)
+                    wrapper = functools.partial(hook, weakref.proxy(self))
```
Comment (albanD) [inline torch/nn/modules/module.py:1169]: You will have to use `weakref.ref()` and update the code a little to check that the ref is still valid to fix the CI.
You want to not use partial anymore and just define a local function I think.

**Comment by otaj** [comment]
Btw, does this PR still make sense since there is this RFC? #75287

File: torch/nn/modules/module.py (hunk 1 above)
Comment (albanD) [reply torch/nn/modules/module.py:1169]: Not sure the backward hook have the same problem as they are created on the fly during the forward pass. So there is no ref cycle being created in this case.
Unless the user stores the output of the forward function on `self.` which should be rare enough that we're happy to make a cycle here.

So only the one for the load state dict would be enough.

**Review by jbschlosser — COMMENTED** [review]
This looks correct to me, capturing a weakref rather than a full reference within the wrapped hook func.

Is there any way to test that the refcount is not increased? It'd be nice to have some validation if possible.

**Comment by otaj** [comment]
That sounds like a reasonable way to test it. Should I write such a test myself, or will you as PyTorch devs write some?

**Comment by jbschlosser** [comment]
> That sounds like a reasonable way to test it. Should I write such a test myself, or will you as PyTorch devs write some?

Yes please :) A good place to add the test is just after the other hook testing in `test/test_nn.py`: https://github.com/pytorch/pytorch/blob/fcf38a5812a164abb591dc5d8e6fd4519ab307be/test/test_nn.py#L810

**Comment by otaj** [comment]
The weekend got in the way, but you got a test, that locally passes, in here :) How does it look now, @jbschlosser @albanD

**Comment by otaj** [comment]
Btw, I just now noticed that you tried to guide me where to put that test :) Hope the current location is ok, if not, I can easily move it around.

File: test/test_nn.py [hunk 2]
```diff
@@ -21263,6 +21265,21 @@ def hook_with_module(module, state_dict, prefix, local_metadata, strict, missing
...
+
+        def hook_with_module(*args, **kwargs):
+            pass
+
+        m._register_load_state_dict_pre_hook(hook_with_module, True)
+        weak_m = weakref.ref(m)
+        del m
+
+        self.assertEqual(weak_m(), None)
+        gc.enable()
```
Comment (albanD) [inline test/test_nn.py:21280]: nit:
```python
try:
  gc.disable()
  # rest of the test
finally:
 gc.enable()
```
We don't want a failure in this test to disable the gc for all other tests!

**Review by albanD — COMMENTED** [review]
The new test looks good.
I think there was an issue with the rebase though as the PR contains a few unrelated changes.
If you can remove these, that should be ready to merge!

File: test/test_nn.py (hunk 2 above)
Comment (otaj) [reply test/test_nn.py]: That's a very good nit! I just pushed a change doing exactly this

**Comment by otaj** [comment]
> The new test looks good. I think there was an issue with the rebase though as the PR contains a few unrelated changes. If you can remove these, that should be ready to merge!

Damn, I really need to step up my git-fu. Anyway, I think I reverted those so all should be good now

**Comment by albanD** [comment]
The diff looks good indeed, but for some reason, github thinks there are some conflicts haha
I think you want to squash all the commits in the branch and force push here.

If you're happy with me pushing in your branch, I can do that for you. Let me know!

**Comment by albanD** [comment]
Sure!
It is a tricky one indeed :p

```
# Get latest pytorch master
git co master
git pull
# Get your branch
git remote add otaj git@github.com:otaj/pytorch.git
git fetch otaj
git co bugfix/weakref
# Do interactive rebase to both get the latest master and be able to remove commits while rebasing
git rebase -i master
# In the prompt, remove (by deleting the corresponding line) the revert commits that shouldn't be here
# Also removed the first 5 commits which seems to be duplicated twice in this branch
git diff master
# Check that everything is here
git push -f
# Force push the new branch to the PR that is now on top of master and without the offending commits
```

**Review by jbschlosser — APPROVED** [review]
LGTM!

**Comment by otaj** [comment]
@albanD Awesome, thanks!

**Comment by malfet** [comment]
Ok, this change is getting reverted as `weak_ref` is not pickable, as one can clearly see by running something as trivial as:
```
% python -c "import torch;import weakref;import pickle; x=torch.rand(3, 3);print(pickle.dumps(x));y=weakref.ref(x);print(pickle.dumps(y))"
```

**Comment by jbschlosser** [comment]
@pytorchbot revert this please

`weakref`s and closure hooks are not pickleable

**Comment by albanD** [comment]
@otaj sorry this got reverted. But even though we absolutely don't recommend it, some users are serializing whole `nn.Module`s.
I think a workaround to this here is going to use a custom python object that will hold the weakref and will have a `__getstate__`, `__setstate__` pair. These will simply take a strong ref to the module when serializing. And convert it back to a weak ref when deserializing. This should solve the problem without causing a cycle again. WDYT?

**Comment by albanD** [comment]
@otaj that sounds good. Opened https://github.com/pytorch/pytorch/issues/78580 to make sure we don't forget and in case someone has time to do that before you.

**Pull request closed without merging**